Two-factors authentication

babolivier opened the pull request:
Activation and configuration panel for the two-factors authentication. Directly follows up https://github.com/cozy/cozy-proxy/pull/260.

This pull request adds a panel to the Settings page under the password change one.
![capture d ecran de 2016-04-15 20-36-58](https://cloud.githubusercontent.com/assets/5547783/14571515/ed669f5a-0349-11e6-9b95-9cb8ff43325f.png)
The user selects the authentication strategy (algorithm) of its choice between the available ones (only HOTP and TOTP for now), then clicks the button. The server will then edit the current user in the DS: it will generate an OTP key (10-char-long random string) and will fill the auth type accordingly ("hotp" or "totp").

Once the user has clicked the button, the page will display a success message and reload after 2s. The Settings page now displays this as the 2FA panel:
![capture d ecran de 2016-04-15 20-46-45](https://cloud.githubusercontent.com/assets/5547783/14571793/73d72df6-034b-11e6-94da-4e1ac57783f1.png)

When the home loads, it checks wether 2FA is enabled to know which panel it has to display. It also retrieve the 2FA token the user will have to enter into its app or device, as long as the corresponding QR code.

The configuration panel for HOTP is a bit different. Indeed, there's an issue we can have with the fact that we store the highest counter encountered until now in order to invalidate any previously used authentication code: when the user switches from authentication app or device to another, its generation will start all the way back to the very first code, with counter 0 (because the new app or device has no way to know the counter we have in the DS nor the corresponding code), which means that, if an user used more than a thousand codes on the old app/device, it'll have to generate as much codes to be able to log back into Cozy.
To address that matter, I've added a button that only shows when HOTP is enabled, which resets the HOTP counter:
![capture d ecran de 2016-04-15 20-53-08](https://cloud.githubusercontent.com/assets/5547783/14572489/63955090-034f-11e6-94f6-6ed9a22efca1.png)
(We can also see in the last two screenshots that the OTP key is regenerated each time we disable/enable back 2FA)

At the time of this writing, I have the process fully working. I now have to:

* Add some tests

Once all of this is done, I'll remove the [WIP] tag.

If you have any question, or anything to say about the PR, what it brings, and the process, please let me know.

Also, @frankrousseau, do you want me to copy all of the above explanation in a doc folder like I did with the proxy?

Edit: Travis is failing, but it doesn't seem that I caused it (the error comes up on the `npm install` command), can I have a hint on that?

frankrousseau commented:
@babolivier Thank you for this second step. If you can add this in a doc folder, it would be nice. If you can post it on the forum as a tech topic, it would be nice too.

babolivier commented:
@frankrousseau Sure thing, will do as soon as I have a stable codebase.

frankrousseau commented on client/app/locales/en.json:
```diff
@@ -98,6 +98,23 @@
   "account identifiers": "Account",
   "account personalization": "Customization",
   "account password": "Password",
+  "account two factors auth": "Two-factors authentication",
```
The right name is "Tow-factor authentication".

https://en.wikipedia.org/wiki/Two-factor_authentication

kosssi replied:
"Two-factor authentication" ;)

frankrousseau commented on client/app/views/account.coffee:
```diff
@@ -108,7 +129,116 @@ module.exports = class exports.AccountView extends BaseView
+        
+        request.post 'api/user', form, (err, data) =>
+            @twoFactorsSubmitButton.spin false
+            if err
```
Maybe you could log the error in the console.

frankrousseau commented on client/app/views/account.coffee:
```diff
@@ -108,7 +129,116 @@ module.exports = class exports.AccountView extends BaseView
+                    setTimeout =>
+                        @error2faAlert.fadeOut()
+                    , 5000
+
```
Put two blank lines between each functions: https://github.com/cozy/cozy-guidelines

frankrousseau commented on client/app/views/account.coffee:
```diff
@@ -108,7 +129,116 @@ module.exports = class exports.AccountView extends BaseView
+    on2faEnableSubmit: (event) ->
+        authType = @twoFactorsField.val()
+        hideFunc = null
+        showSuccess = (message) =>
```
It could be in a separated functio.

frankrousseau commented on client/app/views/account.coffee:
```diff
@@ -108,7 +129,116 @@ module.exports = class exports.AccountView extends BaseView
+        form =
+            authType: authType
+            otpKey: @getOtpKey()
+            hotpCounter: 0 if authType is 'hotp'
```
Are you sure you want to reset the counter each time there is a submission?

babolivier replied:
Yes, as a new OTP key is generated at each submission (and overrides an old one if present). I think we must suppose each activation (or deactivation followed by activation) of OTP will need the user to reconfigure its 2FA device or app like the first time, for more safety.

frankrousseau commented on package.json:
```diff
@@ -28,10 +28,12 @@
     "multiparty": "4.1.2",
     "node-polyglot": "0.4.1",
     "printit": "0.1.18",
+    "qrcode": "^0.4.2",
```
Please, don't add `^` to versions.

frankrousseau commented on server/controllers/account.coffee:
```diff
@@ -97,11 +106,49 @@ module.exports =
+                return res.status(400).send error: err
+            else if users.length is 0
+                return res.status(400).send 
+                    error: localizationManager.t "no user registered"
```
I think you can add a else below this line.

frankrousseau commented on server/controllers/account.coffee:
```diff
@@ -97,11 +106,49 @@ module.exports =
+            token = base32.encode(user.otpKey).toString()
+            provider = "Cozy"
+            twoFAUrl = "otpauth://" + authType
+            twoFAUrl += "/" + provider + ":" + email + "?secret=" + token
```
Please, make coffeescript concatenation.

frankrousseau commented on server/models/user.coffee:
```diff
@@ -7,6 +7,9 @@ module.exports = User = cozydb.getModel 'User',
     password: String
     owner: {type: Boolean, default: false}
     activated: {type: Boolean, default: false}
+    authType: String,
```
Don't add comma here.

babolivier commented:
OK so after discussing with @m4dz, I changed the way the QR Code is being generated. Instead a heavy Node module ran server-side, we now generate the QR Code client-side via [qr.js](https://github.com/neocotic/qr.js), which fixes the problem with the tests not running, and allowed me to remove a couple of functions and a server route.

I've also added the doc in a separate `.md` file.

@frankrousseau, can you please tell me if there's still anything that troubles you?

frankrousseau commented:
Yes the WIP tag. Should we still consider it's a work in progress PR or could we proceed to the functional review?

m4dz commented on client/app/views/account.coffee:
```diff
@@ -108,7 +131,122 @@ module.exports = class exports.AccountView extends BaseView
+    # OTP key is generated client-side, in this function. If consists in a
+    # 10-char long alphanumeric random string.
+    getOtpKey: () ->
+        buf = []
```
We probably should prefer a PRNG rather than the `Math.random()` method to generate a cryptographically randomized string. To generate a hex representation of a 128bits hash (aka a 32chars hex string):
```coffee
# 2 chars per byte, 8 chars per PRNG, so 4 PRNG for a 128bits lengths
PRNG = new Uint32Array 4
window.crypto.getRandomValues PRNG

key = PRNG.reduce (key, subkey) ->
    # pad each PRNG hex with '0'
    key += ('0000' + subkey.toString(16)).slice(-8)
, ''
```

m4dz commented:
I think we faced a potential security leak with the secret key: according to the [RFC](https://tools.ietf.org/html/rfc6238), it should be stored in a cryptographic manner and be deciphered only to check token validity, then re-encrypted immediately after to avoid RAM compromission.

I suggest we store the key the same manner than we do for other sensitive data: using the user password. This way, we decipher the key to check the token using the prompted password at auth, then flush the deciphered password immediately.

@babolivier, any though about this strategy?

babolivier commented:
@frankrousseau The WIP tag is still here only because I haven't added unit tests yet. All the rest can be reviewed. But if you prefer to wait for them in order to review all at once, I'm fine with this.

@m4dz I must admit I have troubles visualizing the concrete implementation, and I think we should have a talk about that. However, I always agree with the highest security :)

frankrousseau commented:
@babolivier are you still working on it? Do you want we finish the job?

babolivier commented:
Hi @frankrousseau,

I've just put this on stand by until the PR on the DS is merged, in order to give some time to my non-profit. But I didn't give up the project and will get back on it soon.

frankrousseau commented:
Ok I merged the PR you mentioned. Thank you again for what you do for the security of Cozy!

babolivier commented:
Pleasure's all mine!
I got the notification about the DS PR. I'll get back on this one tonight (if possible, if not, tomorrow), and hopefully will have everything working here before the end of the week.

babolivier commented:
@frankrousseau Finally, I consider the PR complete and ready, tests are written and green, and have removed the WIP tag! Ready for the final review!
